src/Model_ELBDM/CPU_ELBDM/CPU_ELBDMSolver_HJ.cpp

```diff
@@ -0,0 +1,588 @@
+
+//                   3.3 use N_TIME_LEVELS-stages RK-algorithm
+                     for (uint tl = 0; tl < time_level + 1; ++tl) {
+                        De_New += RK_COEFFS[time_level][tl] * s_In[sj][tl][DENS][si];
+                        Ph_New += RK_COEFFS[time_level][tl] * s_In[sj][tl][PHAS][si];
+                     }
+
+//                   3.5 while computing the temporary results in RK algorithm, just write them to s_In
+                     s_In[sj][time_level + 1][DENS][si] = De_New;
+                     s_In[sj][time_level + 1][PHAS][si] = Ph_New;
+                  }
+
+#                 ifdef  __CUDACC__
+                  __syncthreads();
+#                 endif
+
+               } // if ( time_level < N_TIME_LEVELS - 1 ) {
+            } // if ( isCompletelyRefined ) ... else
+
+//          4. write back final results to g_Fluid_In or g_Fluid_Out to save memory
+
+//          4.1 write shared memory array back to global memory
+            CELL_LOOP(HYB_NXT, HYB_GHOST_SIZE, HYB_GHOST_SIZE)
+            {
+
+               j = j_gap + ( sj + Column0 ) % size_j ;
+               k = k_gap + ( sj + Column0 ) / size_j;
+
+               const real Ph_Old = s_In[sj][0][PHAS][si];
+               real De_New       = s_In[sj][N_TIME_LEVELS][DENS][si];
+               real Ph_New       = s_In[sj][N_TIME_LEVELS][PHAS][si];
+
+//             detect failure of fluid scheme
+               if ( s_HasWaveCounterpart[sj][si] && (De_New < 0 || De_New != De_New || Ph_New != Ph_New || FABS(Ph_New - Ph_Old) > M_PI )) {
+//                   compute real and imaginary parts
+                     const real Re_c       = SQRT(s_In[sj][0][DENS][si    ]) * COS(s_In[sj][0][PHAS][si    ]);
+                     const real Re_p1      = SQRT(s_In[sj][0][DENS][si + 1]) * COS(s_In[sj][0][PHAS][si + 1]);
+                     const real Re_m1      = SQRT(s_In[sj][0][DENS][si - 1]) * COS(s_In[sj][0][PHAS][si - 1]);
+                     const real Im_c       = SQRT(s_In[sj][0][DENS][si    ]) * SIN(s_In[sj][0][PHAS][si    ]);
+                     const real Im_p1      = SQRT(s_In[sj][0][DENS][si + 1]) * SIN(s_In[sj][0][PHAS][si + 1]);
+                     const real Im_m1      = SQRT(s_In[sj][0][DENS][si - 1]) * SIN(s_In[sj][0][PHAS][si - 1]);
+//                   compute laplacians
+                     const real Re_Lap     = Re_p1 - 2 * Re_c + Re_m1;
+                     const real Im_Lap     = Im_p1 - 2 * Im_c + Im_m1;
+//                   second-order centered in space forward in time wave equation update
+                     const real Re_New     = Re_c - Coeff2 * Im_Lap;
+                     const real Im_New     = Im_c + Coeff2 * Re_Lap;
+//                   convert back to phase, match to old phase and then subtract it to get dPhase with NewPhase = OldPhase + dPhase
+                     Ph_New                = EQUALISE( Ph_Old, SATAN2(Im_New, Re_New) );
+                     De_New                = SQR(Re_New) + SQR(Im_New);
```

Comment:
[Important] This will cause inconsistency between the output density and output flux (i.e., `g_Flux[]`). Especially, I notice a case where `De_New` becomes NaN (likely due to negative input density in `g_Fluid_In[]`) and is then fixed by the above "detect failure of fluid scheme" procedure. However, the output flux remains to be NaN, which later on causes problems during the flux fix-up operation.